dialer.go

```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
-		numLive:         1, // the nullSession
-		emaRTT:          ema.NewDuration(0, 0.5),
+	var lc ClientLifecycleListener
+	if opts.Lifecycle != nil {
+		lc = opts.Lifecycle
+	} else {
+		lc = NoopClientLifecycleListener()
+	}
+	d := &dialer{
+		windowSize:        opts.WindowSize,
+		maxPadding:        opts.MaxPadding,
+		maxStreamsPerConn: opts.MaxStreamsPerConn,
+		pingInterval:      opts.PingInterval,
+		pool:              opts.Pool,
+		cipherCode:        opts.Cipher,
+		serverPublicKey:   opts.ServerPublicKey,
+		liveSessions:      make(chan sessionIntf, opts.LiveConns),
+		emaRTT:            ema.NewDuration(0, 0.5),
+		dial:              opts.Dial,
+		pendingSessions:   make(chan *pendingSession, opts.LiveConns),
+		lifecyle:          lc,
+		name:              opts.Name,
+	}
+	d.lifecyle.OnStart()
+	for i := 0; i < opts.LiveConns-1; i++ {
+		d.pendingSessions <- newPendingSession(d.name)
+	}
+
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &pendingSession{
+		name:         "liveness to " + d.name,
+		dialTimeout:  5 * time.Second,
+		sleepOnError: 1 * time.Second,
 	}
+	go d.maintainTCPConnections()
+	return d
 }
 
 type dialer struct {
-	windowSize            int
-	maxPadding            int
-	maxLiveConns          int
-	maxStreamsPerConn     uint16
-	idleInterval          time.Duration
-	pingInterval          time.Duration
-	redialSessionInterval time.Duration
-	pool                  BufferPool
-	cipherCode            Cipher
-	serverPublicKey       *rsa.PublicKey
-	muNumLivePending      sync.Mutex
-	numLive               int
-	numPending            int
-	liveSessions          chan sessionIntf
-	emaRTT                *ema.EMA
+	windowSize        int
+	maxPadding        int
+	maxStreamsPerConn uint16
+	pingInterval      time.Duration
+	pool              BufferPool
+	cipherCode        Cipher
+	serverPublicKey   *rsa.PublicKey
+	liveSessions      chan sessionIntf
+	pendingSessions   chan *pendingSession
+	emaRTT            *ema.EMA
+	dial              DialFN
+	lifecyle          ClientLifecycleListener
+	name              string
 }
 
-func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialContext(context.Background(), dial)
+// maintainTCPConnections maintains background TCP connection(s) and associated lampshade session(s)
+func (d *dialer) maintainTCPConnections() {
+	for rs := range d.pendingSessions {
+		go d.trySession(rs)
+	}
 }
 
-func (d *dialer) DialContext(ctx context.Context, dial DialFN) (net.Conn, error) {
-	s, err := d.getOrCreateSession(ctx, dial)
+func (d *dialer) trySession(ps *pendingSession) {
+	start := time.Now()
+	s, err := d.startSession(ps)
 	if err != nil {
-		return nil, err
+		log.Debugf("Error starting session '%v': %v", ps.name, err.Error())
+		time.Sleep(ps.sleepOnError)
+		d.pendingSessions <- ps
+	} else {
+		log.Debugf("Created session in %v to %#v", time.Since(start), ps)
+		d.liveSessions <- s
```

Comment:
Hmm right now when my Lantern resumes from disconnected network a lot of goroutines are blocked here. How did these requests sneak into `pendingSessions` in the first place?

```
goroutine profile: total 651
289 @ 0x40324ff 0x4007a53 0x4007a29 0x40077a5 0x4740db9 0x4061591
#	0x4740db8	github.com/getlantern/lampshade.(*dialer).trySession+0x418	/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190904225539-ee3e36398b21/dialer.go:148

76 @ 0x40324ff 0x404309b 0x4728c3f 0x4061591
#	0x4728c3e	github.com/getlantern/idletiming.Conn.func1+0x1fe	/Users/joesis/go/pkg/mod/github.com/getlantern/idletiming@v0.0.0-20190529182719-d2fbc83372a5/idletiming_conn.go:64
```

Reply:
```
goroutine profile: total 2003
1669 @ 0x40324ff 0x4007a53 0x4007a29 0x40077a5 0x4740db9 0x4061591
#	0x4740db8	github.com/getlantern/lampshade.(*dialer).trySession+0x418	/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190904225539-ee3e36398b21/dialer.go:148
```

Reply:
Okay here's what happened:
1.  When closing a lampshade session, it won't be removed from `d.liveSessions` until next `d.DialContext`, which means:
1. If the dialer is never selected again by balancer, `d.liveSessions` will never have room for the newly created session, so `trySession` blocks on sending to `d.liveSessions`.
1. The new session can also have error or be idle closed, which triggers yet another new session, and again, blocks on sending to `d.liveSessions`.
1. The same happens over and over again, causes tons of goroutines waiting to add already closed sessions to `d.liveSessions`.

A naive solution would be looping over `d.liveSessions` to remove any closed ones before adding the new one, but there are definitely better ways.

These rounds of bug hunting also made me wondering if the current logic is solid enough so we can just improve over it?

BTW `s.onSessionError` can be called for one session several times, hence `s.Close()`, that may be why requesting new session in `s.Close()` didn't work. If `s.Close()` is protected by `sync.Once`, the closing logic can be moved there.

Reply:
Ah great catch again @joesis! I still think this approach is significantly easier to work with in the end. I added a first pass at a solution but not quite ready for review yet!

Reply:
Gonna play with making it a state machine

Reply:
OK I just added a `recycleSession` method that should fix that particular issue @joesis. I'm super tempted to just turn the whole sequence into a state machine, but that's a bigger change. Might just leave it for now.